_posts/2018-06-12-making-cloudformation-awesome.md

```diff
@@ -0,0 +1,238 @@
+
+Consider an example, in which Acme Corporation has three microservices:
+
+* https://products.acme.com
+* https://orders.acme.com
+* https://customers.acme.com
+
+All services are kept in separate git repositories (`acme-products-api`, `acme-orders-api` and `acme-customers-api`), but because of the cost optimisation policy, all of these services must be served via one load balancer using host based routing, and all docker containers must be deployed in the same ECS cluster.
+
+In a scenario like this, it's a pretty good idea to spin up a new stack `acme-infra` for defining shared resources. So setup would be something similar to this.
+
+```
+acme-infra
+    acme-infra.yml (vpc, SecurityGroups, ECS cluster + roles, ALB)
+acme-products-api
+    acme-products-service.yml (ECS definitions, target group, ALB routing rules, Imports necessary stuff)
+acme-orders-api
+    acme-orders-service.yml (ECS definitions, target group, ALB routing rules, Imports necessary stuff)
+acme-customers-api
+    acme-customers-service.yml (ECS definitions, target group, ALB routing rules, Imports necessary stuff)    
+```
+
+Using shared ALB like this requires some additional tweaks, i.e sharing routing priority numbers between services, but that is an implementation detail. Rule of the thumb is that you should have just the right amount of stacks and dependency graph must be unidirectional. So kids! Remember not to use `!Import` in the infra stack! 😘
+
+It's super important to understand in which stack you should add your new resource in a multi-stack scenario, because you can really easily make circular dependencies between stacks. Think for instance if you have created separate stacks for [CloudFront](https://aws.amazon.com/cloudfront/) and [S3 buckets](https://aws.amazon.com/s3/). Your CloudFront stack needs to know S3 bucket domains and your S3 bucket policy is S3 stack needs to know CloudFront [Origin Access Identity](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-restricting-access-to-s3.html). There you have a circular dependency and you probably spend loadsa time debugging and possibly even suffer some downtime when stuff hits the fan.
+
+But the good news is that when you just bake both CloudFront and S3 stuff into the same stack - it just works.
+
+Bottom line is that, when having multiple stacks, you should be really careful when selecting a stack for your new resource. It's a good idea to have well maintained visualisation of the stacks and exports/imports. Don't be afraid to make a loose coupling between resources when needed! It's totally ok to just copy resource references directly as a parameters to other stacks instead of exporting and importing them. Exporting is a really powerful tool and it makes really hard coupling between the stacks. Try to minimise the amount of stacks. Less is more.
```

Comment:
to have _a_ well maintained visual(isation)

directly as parameters (ylimääräinen 'a' pois)

mulle on epäselvä tän merkitys:  "it makes really hard coupling between the stacks"